python/ray/data/_internal/compute.py

```diff
@@ -254,15 +254,36 @@ def _apply(
 
         if name is None:
             name = "map"
-        blocks_in = block_list.get_blocks_with_metadata()
-        # Bin blocks by target block size.
+        blocks_in: List[
+            Tuple[ObjectRef[Block], BlockMetadata]
+        ] = block_list.get_blocks_with_metadata()
+
+        # We bundle blocks according to the following rules:
+        # 1. Attempt to bundle up to the target block size.
+        # 2. If the max concurrency of the ActorPool is set, then
+        #    cap the number of bundles to match the size of the ActorPool.
+        #    This avoids additional overhead in submitting new actor tasks and allows
+        #    the actor task to do optimizations such as batch prefetching.
+        target_size = -1
+        # First bundle up to the provided target block size.
         if target_block_size is not None:
             _check_batch_size(blocks_in, target_block_size, name)
-            block_bundles = _bundle_blocks_up_to_size(
-                blocks_in, target_block_size, name
+            target_size = max(target_block_size, target_size)
+        # If the max size of the actor pool is set, then we bundle up even more
+        # if necessary.
+        if not math.isinf(self.max_size):
+            total_size = sum(
+                metadata.num_rows if metadata.num_rows is not None else float("inf")
```

Comment:
Falling back to `float("inf")` would make the `target_size` infinite, which means all blocks would be put into a single bundle, which I don't think is what we'd want.

Reply:
updated and added test